File: mock-src/python/tasktracker/services/user_service.py
"""User service for managing users."""

import asyncio
import hashlib
import secrets
from typing import Optional
from uuid import UUID

from ..models.user import User, UserRole
from ..repositories.user_repository import UserRepository
from ..utils.validators import validate_email, validate_username, validate_password
from .base import (
    BaseService,
    NotFoundError,
    ValidationError,
    AuthorizationError,
    log_call,
)


class UserService(BaseService):
    """
    Service for user management operations.

    Handles user creation, authentication, and authorization.

    Attributes:
        repository: The user repository for data access.
    """

    def __init__(self, repository: Optional[UserRepository] = None) -> None:
        """
        Initialize the user service.

        Args:
            repository: Optional repository instance.
        """
        super().__init__()
        self.repository = repository or UserRepository()
        self._password_hashes: dict[UUID, str] = {}
        self._session_tokens: dict[str, UUID] = {}

    @log_call
    def create_user(
        self,
        username: str,
        email: str,
        password: str,
        display_name: str = "",
        role: UserRole = UserRole.MEMBER,
    ) -> User:
        """
        Create a new user account.

        Args:
            username: Unique username for login.
            email: User's email address.
            password: Account password.
            display_name: Name to display in UI.
            role: User's access level.

        Returns:
            The created user.

        Raises:
            ValidationError: If any field is invalid.
        """
        # Validate inputs
        if not validate_username(username):
            raise ValidationError("username", "Invalid username format")

        if not validate_email(email):
            raise ValidationError("email", "Invalid email format")

        is_valid, errors = validate_password(password)
        if not is_valid:
            raise ValidationError("password", "; ".join(errors))

        # Check uniqueness
        if self.repository.email_exists(email):
            raise ValidationError("email", "Email already registered")

        if self.repository.username_exists(username):
            raise ValidationError("username", "Username already taken")

        # Create user
        user = User(
            username=username,
            email=email,
            display_name=display_name or username,
            role=role,
        )

        created = self.repository.create(user)

        # Store password hash
        self._password_hashes[created.id] = self._hash_password(password)

        self._log_info("Created user: %s", created.username)
        return created

    @log_call
    def get_user(self, user_id: UUID) -> User:
        """
        Get a user by ID.

        Args:
            user_id: The user's unique identifier.

        Returns:
            The user.

        Raises:
            NotFoundError: If user doesn't exist.
        """
        user = self.repository.get(user_id)
        if not user:
            raise NotFoundError("User", str(user_id))
        return user

    def get_user_by_email(self, email: str) -> User:
        """
        Get a user by email address.

        Args:
            email: The email to look up.

        Returns:
            The user.

        Raises:
            NotFoundError: If user doesn't exist.
        """
        user = self.repository.get_by_email(email)
        if not user:
            raise NotFoundError("User", email)
        return user

    def get_user_by_username(self, username: str) -> User:
        """
        Get a user by username.

        Args:
            username: The username to look up.

        Returns:
            The user.

        Raises:
            NotFoundError: If user doesn't exist.
        """
        user = self.repository.get_by_username(username)
        if not user:
            raise NotFoundError("User", username)
        return user

    @log_call
    def update_user(
        self,
        user_id: UUID,
        display_name: Optional[str] = None,
        email: Optional[str] = None,
    ) -> User:
        """
        Update a user's profile.

        Args:
            user_id: The user to update.
            display_name: New display name.
            email: New email address.

        Returns:
            The updated user.
        """
        user = self.get_user(user_id)

        if display_name is not None:
            user.display_name = display_name

        if email is not None:
            if not validate_email(email):
                raise ValidationError("email", "Invalid email format")
            # Note: In real impl, would update email index
            user.email = email

        self.repository.update(user)
        return user

    @log_call
    def deactivate_user(self, user_id: UUID, actor_id: UUID) -> User:
        """
        Deactivate a user account.

        Args:
            user_id: The user to deactivate.
            actor_id: The user performing the action.

        Returns:
            The deactivated user.

        Raises:
            AuthorizationError: If actor lacks permission.
        """
        actor = self.get_user(actor_id)
        if not actor.is_admin:
            raise AuthorizationError("deactivate", "user accounts")

        user = self.get_user(user_id)
        user.deactivate()
        self.repository.update(user)
        return user

    def authenticate(self, username: str, password: str) -> Optional[str]:
        """
        Authenticate a user and return a session token.

        Args:
            username: The username to authenticate.
            password: The password to verify.

        Returns:
            Session token if successful, None if failed.
        """
        try:
            user = self.get_user_by_username(username)
        except NotFoundError:
            return None

        if not user.is_active:
            return None

        stored_hash = self._password_hashes.get(user.id)
        if not stored_hash:
            return None

        if self._hash_password(password) != stored_hash:
            return None

        # Create session token
        token = self._generate_token()
        self._session_tokens[token] = user.id

        # Record login
        user.record_login()
        self.repository.update(user)

        return token

    async def authenticate_async(
        self,
        username: str,
        password: str,
    ) -> Optional[str]:
        """
        Authenticate a user asynchronously.

        Args:
            username: The username to authenticate.
            password: The password to verify.

        Returns:
            Session token if successful, None if failed.
        """
        await asyncio.sleep(0.01)  # Simulate async verification
        return self.authenticate(username, password)

    def validate_token(self, token: str) -> Optional[User]:
        """
        Validate a session token and return the user.

        Args:
            token: The session token to validate.

        Returns:
            The user if valid, None otherwise.
        """
        user_id = self._session_tokens.get(token)
        if not user_id:
            return None

        try:
            user = self.get_user(user_id)
            return user if user.is_active else None
        except NotFoundError:
            return None

    def logout(self, token: str) -> bool:
        """
        Invalidate a session token.

        Args:
            token: The token to invalidate.

        Returns:
            True if token was valid and invalidated.
        """
        if token in self._session_tokens:
            del self._session_tokens[token]
            return True
        return False

    def change_password(
        self,
        user_id: UUID,
        old_password: str,
        new_password: str,
    ) -> bool:
        """
        Change a user's password.

        Args:
            user_id: The user's ID.
            old_password: Current password for verification.
            new_password: New password to set.

        Returns:
            True if password was changed.

        Raises:
            ValidationError: If new password is invalid.
            AuthorizationError: If old password is wrong.
        """
        user = self.get_user(user_id)

        stored_hash = self._password_hashes.get(user.id)
        if self._hash_password(old_password) != stored_hash:
            raise AuthorizationError("change", "password")

        is_valid, errors = validate_password(new_password)
        if not is_valid:
            raise ValidationError("password", "; ".join(errors))

        self._password_hashes[user.id] = self._hash_password(new_password)
        return True

    def promote_user(
        self,
        user_id: UUID,
        new_role: UserRole,
        actor_id: UUID,
    ) -> User:
        """
        Promote a user to a higher role.

        Args:
            user_id: The user to promote.
            new_role: The new role to assign.
            actor_id: The user performing the promotion.

        Returns:
            The updated user.

        Raises:
            AuthorizationError: If actor lacks permission.
        """
        actor = self.get_user(actor_id)
        if not actor.is_admin:
            raise AuthorizationError("promote", "users")

        user = self.get_user(user_id)
        user.promote_to(new_role)
        self.repository.update(user)
        return user

    def get_admins(self) -> list[User]:
        """Get all admin users."""
        return self.repository.get_admins()

    def search_users(self, query: str) -> list[User]:
        """
        Search users by name, email, or username.

        Args:
            query: Search text.

        Returns:
            Matching users.
        """
        return self.repository.search(query)

    @staticmethod
    def _hash_password(password: str) -> str:
        """Hash a password using SHA-256."""
        return hashlib.sha256(password.encode()).hexdigest()

    @staticmethod
    def _generate_token() -> str:
        """Generate a secure session token."""
        return secrets.token_urlsafe(32)

    @property
    def active_user_count(self) -> int:
        """Get count of active users."""
        return len(self.repository.get_active())

File: mock-src/typescript/src/services/UserService.ts
/**
 * User service for managing users.
 */

import { User, UserRole, IUser } from '../models/User';
import {
  validateEmail,
  validateUsername,
  validatePassword,
} from '../utils/validators';
import { generateId } from '../utils/helpers';

/**
 * Service error base class.
 */
export class ServiceError extends Error {
  constructor(
    message: string,
    public code: string = 'SERVICE_ERROR'
  ) {
    super(message);
    this.name = 'ServiceError';
  }
}

/**
 * Not found error.
 */
export class NotFoundError extends ServiceError {
  constructor(entityType: string, entityId: string) {
    super(`${entityType} with ID ${entityId} not found`, 'NOT_FOUND');
    this.name = 'NotFoundError';
  }
}

/**
 * Validation error.
 */
export class ValidationError extends ServiceError {
  constructor(
    public field: string,
    message: string
  ) {
    super(`Validation failed for ${field}: ${message}`, 'VALIDATION_ERROR');
    this.name = 'ValidationError';
  }
}

/**
 * Authorization error.
 */
export class AuthorizationError extends ServiceError {
  constructor(action: string, resource: string) {
    super(`Not authorized to ${action} ${resource}`, 'AUTHORIZATION_ERROR');
    this.name = 'AuthorizationError';
  }
}

/**
 * Options for creating a user.
 */
export interface CreateUserOptions {
  displayName?: string;
  role?: UserRole;
}

/**
 * Service for user management operations.
 *
 * Handles user creation, authentication, and authorization.
 */
export class UserService {
  private users: Map<string, User> = new Map();
  private emailIndex: Map<string, string> = new Map();
  private usernameIndex: Map<string, string> = new Map();
  private passwordHashes: Map<string, string> = new Map();
  private sessionTokens: Map<string, string> = new Map();

  /**
   * Create a new user account.
   *
   * @param username - Unique username for login
   * @param email - User's email address
   * @param password - Account password
   * @param options - Additional user options
   * @returns The created user
   * @throws ValidationError if any field is invalid
   */
  createUser(
    username: string,
    email: string,
    password: string,
    options: CreateUserOptions = {}
  ): User {
    // Validate inputs
    if (!validateUsername(username)) {
      throw new ValidationError('username', 'Invalid username format');
    }

    if (!validateEmail(email)) {
      throw new ValidationError('email', 'Invalid email format');
    }

    const passwordValidation = validatePassword(password);
    if (!passwordValidation.isValid) {
      throw new ValidationError('password', passwordValidation.errors.join('; '));
    }

    // Check uniqueness
    if (this.emailIndex.has(email.toLowerCase())) {
      throw new ValidationError('email', 'Email already registered');
    }

    if (this.usernameIndex.has(username.toLowerCase())) {
      throw new ValidationError('username', 'Username already taken');
    }

    // Create user
    const user = new User(username, email, {
      displayName: options.displayName || username,
      role: options.role,
    });

    this.users.set(user.id, user);
    this.emailIndex.set(email.toLowerCase(), user.id);
    this.usernameIndex.set(username.toLowerCase(), user.id);
    this.passwordHashes.set(user.id, this.hashPassword(password));

    return user;
  }

  /**
   * Get a user by ID.
   *
   * @param userId - The user's unique identifier
   * @returns The user
   * @throws NotFoundError if user doesn't exist
   */
  getUser(userId: string): User {
    const user = this.users.get(userId);
    if (!user) {
      throw new NotFoundError('User', userId);
    }
    return user;
  }

  /**
   * Get a user by email address.
   *
   * @param email - The email to look up
   * @returns The user
   * @throws NotFoundError if user doesn't exist
   */
  getUserByEmail(email: string): User {
    const userId = this.emailIndex.get(email.toLowerCase());
    if (!userId) {
      throw new NotFoundError('User', email);
    }
    return this.getUser(userId);
  }

  /**
   * Get a user by username.
   *
   * @param username - The username to look up
   * @returns The user
   * @throws NotFoundError if user doesn't exist
   */
  getUserByUsername(username: string): User {
    const userId = this.usernameIndex.get(username.toLowerCase());
    if (!userId) {
      throw new NotFoundError('User', username);
    }
    return this.getUser(userId);
  }

  /**
   * Update a user's profile.
   *
   * @param userId - The user to update
   * @param displayName - New display name
   * @param email - New email address
   * @returns The updated user
   */
  updateUser(
    userId: string,
    displayName?: string,
    email?: string
  ): User {
    const user = this.getUser(userId);

    if (displayName !== undefined) {
      user.displayName = displayName;
    }

    if (email !== undefined) {
      if (!validateEmail(email)) {
        throw new ValidationError('email', 'Invalid email format');
      }
      user.email = email;
    }

    return user;
  }

  /**
   * Deactivate a user account.
   *
   * @param userId - The user to deactivate
   * @param actorId - The user performing the action
   * @returns The deactivated user
   * @throws AuthorizationError if actor lacks permission
   */
  deactivateUser(userId: string, actorId: string): User {
    const actor = this.getUser(actorId);
    if (!actor.isAdmin) {
      throw new AuthorizationError('deactivate', 'user accounts');
    }

    const user = this.getUser(userId);
    user.deactivate();
    return user;
  }

  /**
   * Authenticate a user and return a session token.
   *
   * @param username - The username to authenticate
   * @param password - The password to verify
   * @returns Session token if successful, null if failed
   */
  authenticate(username: string, password: string): string | null {
    let user: User;
    try {
      user = this.getUserByUsername(username);
    } catch {
      return null;
    }

    if (!user.isActive) {
      return null;
    }

    const storedHash = this.passwordHashes.get(user.id);
    if (!storedHash || this.hashPassword(password) !== storedHash) {
      return null;
    }

    // Create session token
    const token = this.generateToken();
    this.sessionTokens.set(token, user.id);

    // Record login
    user.recordLogin();

    return token;
  }

  /**
   * Authenticate a user asynchronously.
   *
   * @param username - The username to authenticate
   * @param password - The password to verify
   * @returns Promise resolving to session token or null
   */
  async authenticateAsync(
    username: string,
    password: string
  ): Promise<string | null> {
    await new Promise((resolve) => setTimeout(resolve, 10));
    return this.authenticate(username, password);
  }

  /**
   * Validate a session token and return the user.
   *
   * @param token - The session token to validate
   * @returns The user if valid, null otherwise
   */
  validateToken(token: string): User | null {
    const userId = this.sessionTokens.get(token);
    if (!userId) {
      return null;
    }

    try {
      const user = this.getUser(userId);
      return user.isActive ? user : null;
    } catch {
      return null;
    }
  }

  /**
   * Invalidate a session token.
   *
   * @param token - The token to invalidate
   * @returns True if token was valid and invalidated
   */
  logout(token: string): boolean {
    if (this.sessionTokens.has(token)) {
      this.sessionTokens.delete(token);
      return true;
    }
    return false;
  }

  /**
   * Change a user's password.
   *
   * @param userId - The user's ID
   * @param oldPassword - Current password for verification
   * @param newPassword - New password to set
   * @returns True if password was changed
   * @throws ValidationError if new password is invalid
   * @throws AuthorizationError if old password is wrong
   */
  changePassword(
    userId: string,
    oldPassword: string,
    newPassword: string
  ): boolean {
    const user = this.getUser(userId);

    const storedHash = this.passwordHashes.get(user.id);
    if (this.hashPassword(oldPassword) !== storedHash) {
      throw new AuthorizationError('change', 'password');
    }

    const passwordValidation = validatePassword(newPassword);
    if (!passwordValidation.isValid) {
      throw new ValidationError('password', passwordValidation.errors.join('; '));
    }

    this.passwordHashes.set(user.id, this.hashPassword(newPassword));
    return true;
  }

  /**
   * Promote a user to a higher role.
   *
   * @param userId - The user to promote
   * @param newRole - The new role to assign
   * @param actorId - The user performing the promotion
   * @returns The updated user
   * @throws AuthorizationError if actor lacks permission
   */
  promoteUser(userId: string, newRole: UserRole, actorId: string): User {
    const actor = this.getUser(actorId);
    if (!actor.isAdmin) {
      throw new AuthorizationError('promote', 'users');
    }

    const user = this.getUser(userId);
    user.promoteTo(newRole);
    return user;
  }

  /**
   * Get all admin users.
   */
  getAdmins(): User[] {
    return Array.from(this.users.values()).filter((u) => u.isAdmin);
  }

  /**
   * Search users by name, email, or username.
   *
   * @param query - Search text
   * @returns Matching users
   */
  searchUsers(query: string): User[] {
    const queryLower = query.toLowerCase();
    return Array.from(this.users.values()).filter(
      (u) =>
        u.username.toLowerCase().includes(queryLower) ||
        u.email.toLowerCase().includes(queryLower) ||
        u.displayName.toLowerCase().includes(queryLower)
    );
  }

  /**
   * Get count of active users.
   */
  get activeUserCount(): number {
    return Array.from(this.users.values()).filter((u) => u.isActive).length;
  }

  private hashPassword(password: string): string {
    // Simple hash for demo purposes
    let hash = 0;
    for (let i = 0; i < password.length; i++) {
      const char = password.charCodeAt(i);
      hash = (hash << 5) - hash + char;
      hash = hash & hash;
    }
    return hash.toString(16);
  }

  private generateToken(): string {
    return generateId() + generateId();
  }
}
